[JIT] parse prim::Constant[value=annotate()] and prim::Constant[value={0}]

davidberard98 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #76875

irparser previously didn't support these, which would cause failures in
log_extract.py

Differential Revision: [D36156699](https://our.internmc.facebook.com/intern/diff/D36156699)

davidberard98 committed:
[JIT] parse prim::Constant[value=annotate()] and prim::Constant[value={0}]

irparser previously didn't support these, which would cause failures in
log_extract.py

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] parse prim::Constant[value=annotate()] and prim::Constant[value={0}]"


irparser previously didn't support these, which would cause failures in
log_extract.py

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156699).

davidberard98 committed:
Update on "[JIT] parse prim::Constant[value=annotate()] and prim::Constant[value={0}]"


irparser previously didn't support these, which would cause failures in
log_extract.py

Differential Revision: [D36156699](https://our.internmc.facebook.com/intern/diff/D36156699)

[ghstack-poisoned]

davidberard98 committed:
Update on "[JIT] parse prim::Constant[value=annotate()] and prim::Constant[value={0}]"


irparser previously didn't support these, which would cause failures in
log_extract.py

Differential Revision: [D36156699](https://our.internmc.facebook.com/intern/diff/D36156699)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156699).

eellison commented on test/jit/test_misc.py:
```diff
@@ -345,3 +345,22 @@ def sum_f(x: BroadcastingList2[float]) -> float:
+        graph = torch._C.parse_ir(ir, True)
+        func = torch._C._create_function_from_graph("forward", graph)
+        ret = func()
+        self.assertTrue(ret.numel() == 1)
```
can you also check that `len(ret.size() == 0)`

davidberard98 replied:
actually it's ret.size() == 1 for this tensor - added a check

eellison commented on test/jit/test_misc.py:
```diff
@@ -345,3 +345,22 @@ def sum_f(x: BroadcastingList2[float]) -> float:
+          %3 : int[] = prim::Constant[value=annotate(List[int], [])]()
+          return (%3)
+        """
+        graph = torch._C.parse_ir(ir, True)
```
nit, check result somehow ? maybe with 
```
%1 : bool = prim::isinstance[types=[int[]]](%3)
return (%1)
```

davidberard98 replied:
hmm that didn't work because... it can't parse List[int] 😄 

I'm just checking `== []` right now, lmk if you have another suggestion on how to do this

eellison commented on test/jit/test_misc.py:
```diff
@@ -345,3 +345,22 @@ def sum_f(x: BroadcastingList2[float]) -> float:
+    def test_parse_ir_single_element_tensor(self):
+        ir = """
+        graph():
+          %7 : Long(1, strides=[1], requires_grad=0, device=cpu) = prim::Constant[value={0}]()
```
nit, could you exercise the negative path as well ?

eellison commented on torch/csrc/jit/ir/irparser.cpp:
```diff
@@ -226,12 +229,41 @@ ParsedLiteral IRParser::parseScalarLiteral(Node* n) {
+      if (L.cur().kind == '-') {
+        L.next();
+      }
+      auto text = L.expect(TK_NUMBER);
```
it would be easy enough to actually recover the exact value here... but I think more consistent to not, as you are currently doing

eellison commented on torch/csrc/jit/ir/irparser.cpp:
```diff
@@ -306,6 +338,29 @@ void IRParser::parseAttr(Node* n) {
+          << "Unexpected annotation (only List and Dict can be parsed)";
+    }
+    L.next();
+    // use the profiling annotation instead of the `annotate()` type.
```
profiling is too specific, since types could come elsewhere. I would say something like `ignore the annotations on the constants IValue, and recover type from the Node output`

eellison commented on torch/csrc/jit/ir/irparser.cpp:
```diff
@@ -306,6 +338,29 @@ void IRParser::parseAttr(Node* n) {
+    // Note: we could also use script_type_parser
+    bypassTypeAnnotationList();
+    L.expect(',');
+    // expect an empty definition
```
It wont always be https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/ivalue.cpp#L484 empty.. although we can defer handling this case

davidberard98 replied:
added this info to the comment

eellison commented on torch/csrc/jit/ir/irparser.cpp:
```diff
@@ -226,12 +229,41 @@ ParsedLiteral IRParser::parseScalarLiteral(Node* n) {
 
+void IRParser::bypassTypeAnnotationList() {
+  int depth = 0;
+  while (!(L.cur().kind == ']' && depth == 1)) {
```
nit: felt the condition here took a little bit to parse. would it be simpler to do something like 
```
bool bypassed_list = false;
int depth = 0;
while (depth != 0 &&  !bypassed_list) {
    if (L.cur().kind == '[') {
      bypassed_list = True;
      depth++;
    } else if (L.cur().kind == ']') {
      depth--;
    }
    L.next();
  }
}
```

eellison commented on torch/csrc/jit/ir/irparser.cpp:
```diff
@@ -545,6 +600,18 @@ void IRParser::parse() {
     (void)t;
   }
+
+  for (Node* n : deferred_empty_container_initializations_) {
```
I don't think we're ever adding to this list

davidberard98 replied:
whoops - thanks for catching that!

eellison reviewed (commented):
looking good, but a couple comments

davidberard98 committed:
Update on "[JIT] parse prim::Constant[value=annotate()] and prim::Constant[value={0}]"


irparser previously didn't support these, which would cause failures in
log_extract.py

Differential Revision: [D36156699](https://our.internmc.facebook.com/intern/diff/D36156699)

[ghstack-poisoned]

davidberard98 commented:
@davidberard98 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36156699).

eellison reviewed (approved):
LGTM

tugsbayasgalan reviewed (approved):
LGTM!

davidberard98 commented:
@pytorchbot merge this please
